### ISSUE TITLE
Screenshot plugin: Tainted canvas

### ISSUE THREAD

**Issue by KTGLeiden** [openseadragon/openseadragon#1043]
Hi there, 
I am finishing up building the openseadragon-screenshot plugin, however I repeatedly get the javascript security error _"openseadragonScreenshot.js:329 -> Uncaught SecurityError: Failed to execute 'toBlob' on 'HTMLCanvasElement': Tainted canvases may not be exported."_
I have never seen this error before, and before I was always using the latest, stable version of OSD (2.2.1). Now, I switched to the latest version in the master branch as this one includes the #837 fix which I will need to implement this. Is this a problem introduced somewhere along the line in the master branch? You can see the problem directly if you go to https://ktgleiden.github.io/test.html. Hope you can help!

Edit: If you want to reproduce the error, please make a screenshot ;) this is done by clicking the button with the square on it.
~Koen

**Comment by avandecreme** [comment]
Hello,
Good work on the screenshot plugin!

Your issue happens because the tiles are hosted on a different host than the one you are hosting the example.
You will need to enable CORS to be able to do screenshots.
Github already has the CORS headers, so all you need to do is open the DZI with the `crossOriginPolicy` set:
```javascript
	    var viewer = OpenSeadragon({
	        id: "openseadragon1",
	        prefixUrl: "images/",
	        tileSources: "https://openseadragon.github.io/example-images/duomo/duomo.dzi",
	        crossOriginPolicy: "Anonymous"
	    });
```

**Comment by KTGLeiden** [comment]
Thx! :)

Awesome! That is something I wasn't aware of. You fixed it right away, which means that it works properly now. I will update the readme.md with this info and add the plugin to the list ;)

~Koen

**Comment by devanonyme** [comment]
I have worked on this today, including finding a user friendly version for IOS (that doesnt really implement user download, user have to click on the image to get the image save dialog (not supported directly by canvas)).

I have a blocking issue with Explorer and it seem to be related to this thread (I was also using toblob / filesaverjs as base solution). The crossOrigin Policy is not sufficient and toBlob (6.https://www.npmjs.com/package/canvas-toBlob) and toDataUrl are blocked.

Is it possible to get it working without using liberal CORS? (* / anonymous). We have control on the serveur CORS but it's currently set to grant right to the serveur on the domain and our local dev machine.

In the other hand, what I have read so far suggest that crossOrigin on IMG is not supported on Explorer.

Someone was able to implement file save on Explorer?

So far, it looks like we will have to implement the solution using the backend with content-disposition and more hoops to make sure the angular app wont be removed on a failed load (I have used jquery file download for this in another project, so I know it's a viable solution).
